avro protocol: key converting to Avro binary failed

### What did you do?

1. There is a table ycsb1.t, row width 4k, with some existing data
```
 CREATE TABLE `t` (
  `YCSB_KEY` varchar(64) NOT NULL,
  `FIELD0` varchar(100) DEFAULT NULL,
  `FIELD1` varchar(100) DEFAULT NULL,
  `FIELD2` varchar(100) DEFAULT NULL,
  `FIELD3` varchar(100) DEFAULT NULL,
  `FIELD4` varchar(100) DEFAULT NULL,
  `FIELD5` varchar(100) DEFAULT NULL,
  `FIELD6` varchar(100) DEFAULT NULL,
  `FIELD7` varchar(100) DEFAULT NULL,
  `FIELD8` varchar(100) DEFAULT NULL,
  `FIELD9` varchar(100) DEFAULT NULL,
  `FIELD10` varchar(100) DEFAULT NULL,
  `FIELD11` varchar(100) DEFAULT NULL,
  `FIELD12` varchar(100) DEFAULT NULL,
  `FIELD13` varchar(100) DEFAULT NULL,
  `FIELD14` varchar(100) DEFAULT NULL,
  `FIELD15` varchar(100) DEFAULT NULL,
  `FIELD16` varchar(100) DEFAULT NULL,
  `FIELD17` varchar(100) DEFAULT NULL,
  `FIELD18` varchar(100) DEFAULT NULL,
  `FIELD19` varchar(100) DEFAULT NULL,
  `FIELD20` varchar(100) DEFAULT NULL,
  `FIELD21` varchar(100) DEFAULT NULL,
  `FIELD22` varchar(100) DEFAULT NULL,
  `FIELD23` varchar(100) DEFAULT NULL,
  `FIELD24` varchar(100) DEFAULT NULL,
  `FIELD25` varchar(100) DEFAULT NULL,
  `FIELD26` varchar(100) DEFAULT NULL,
  `FIELD27` varchar(100) DEFAULT NULL,
  `FIELD28` varchar(100) DEFAULT NULL,
  `FIELD29` varchar(100) DEFAULT NULL,
  `FIELD30` varchar(100) DEFAULT NULL,
  `FIELD31` varchar(100) DEFAULT NULL,
  `FIELD32` varchar(100) DEFAULT NULL,
  `FIELD33` varchar(100) DEFAULT NULL,
  `FIELD34` varchar(100) DEFAULT NULL,
  `FIELD35` varchar(100) DEFAULT NULL,
  `FIELD36` varchar(100) DEFAULT NULL,
  `FIELD37` varchar(100) DEFAULT NULL,
  `FIELD38` varchar(100) DEFAULT NULL,
  `FIELD39` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`YCSB_KEY`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```
3. create avro changefeed
4. run gcsb workload

### What did you expect to see?

changefeed no error.

### What did you see instead?

changefeed error.

![img_v3_028u_2cadba27-58a4-43d1-8b8e-dfd4a7bc4afg](https://github.com/pingcap/tiflow/assets/7403864/7af4e9b5-4620-43a1-8374-aea441321794)

````
[2024/03/13 13:29:57.703 +08:00] [ERROR] [avro.go:102] ["avro: key converting to Avro binary failed"] [error="cannot encode binary record \"default.ycsb1.t\" field \"YCSB_KEY\": schema does not specify default value and no value provided"]
```

### Versions of the cluster

CDC version: v7.5.1

--- a/pkg/sink/codec/avro/avro.go
+++ b/pkg/sink/codec/avro/avro.go
@@ -93,13 +93,15 @@ func (a *BatchEncoder) encodeKey(ctx context.Context, topic string, e *model.Row
 
 	native, err := a.columns2AvroData(keyColumns)
 	if err != nil {
-		log.Error("avro: key converting to native failed", zap.Error(err))
+		log.Error("avro: key converting to native failed",
+			zap.Error(err), zap.Any("columns", cols), zap.Any("colInfos", colInfos))
 		return nil, errors.Trace(err)
 	}
 
 	bin, err := avroCodec.BinaryFromNative(nil, native)
 	if err != nil {
-		log.Error("avro: key converting to Avro binary failed", zap.Error(err))
+		log.Error("avro: key converting to Avro binary failed",
+			zap.Error(err), zap.Any("columns", cols), zap.Any("colInfos", colInfos))
 		return nil, cerror.WrapError(cerror.ErrAvroEncodeToBinary, err)
 	}
 

--- a/pkg/sink/codec/avro/avro.go
+++ b/pkg/sink/codec/avro/avro.go
@@ -180,7 +180,8 @@ func (a *BatchEncoder) encodeValue(ctx context.Context, topic string, e *model.R
 
 	native, err := a.columns2AvroData(input)
 	if err != nil {
-		log.Error("avro: converting value to native failed", zap.Error(err))
+		log.Error("avro: converting value to native failed",
+			zap.Error(err), zap.Any("columns", input.columns), zap.Any("colInfos", input.colInfos))
 		return nil, errors.Trace(err)
 	}
 	if a.config.EnableTiDBExtension {
@@ -189,7 +190,8 @@ func (a *BatchEncoder) encodeValue(ctx context.Context, topic string, e *model.R
 
 	bin, err := avroCodec.BinaryFromNative(nil, native)
 	if err != nil {
-		log.Error("avro: converting value to Avro binary failed", zap.Error(err))
+		log.Error("avro: converting value to Avro binary failed",
+			zap.Error(err), zap.Any("columns", input.columns), zap.Any("colInfos", input.colInfos))
 		return nil, cerror.WrapError(cerror.ErrAvroEncodeToBinary, err)
 	}
 

--- a/pkg/sink/codec/avro/avro.go
+++ b/pkg/sink/codec/avro/avro.go
@@ -93,15 +93,13 @@ func (a *BatchEncoder) encodeKey(ctx context.Context, topic string, e *model.Row
 
 	native, err := a.columns2AvroData(keyColumns)
 	if err != nil {
-		log.Error("avro: key converting to native failed",
-			zap.Error(err), zap.Any("columns", cols), zap.Any("colInfos", colInfos))
+		log.Error("avro: key converting to native failed", zap.Error(err))
 		return nil, errors.Trace(err)
 	}
 
 	bin, err := avroCodec.BinaryFromNative(nil, native)
 	if err != nil {
-		log.Error("avro: key converting to Avro binary failed",
-			zap.Error(err), zap.Any("columns", cols), zap.Any("colInfos", colInfos))
+		log.Error("avro: key converting to Avro binary failed", zap.Error(err))
 		return nil, cerror.WrapError(cerror.ErrAvroEncodeToBinary, err)
 	}
 
@@ -180,8 +178,7 @@ func (a *BatchEncoder) encodeValue(ctx context.Context, topic string, e *model.R
 
 	native, err := a.columns2AvroData(input)
 	if err != nil {
-		log.Error("avro: converting value to native failed",
-			zap.Error(err), zap.Any("columns", input.columns), zap.Any("colInfos", input.colInfos))
+		log.Error("avro: converting value to native failed", zap.Error(err))
 		return nil, errors.Trace(err)
 	}
 	if a.config.EnableTiDBExtension {
@@ -190,8 +187,7 @@ func (a *BatchEncoder) encodeValue(ctx context.Context, topic string, e *model.R
 
 	bin, err := avroCodec.BinaryFromNative(nil, native)
 	if err != nil {
-		log.Error("avro: converting value to Avro binary failed",
-			zap.Error(err), zap.Any("columns", input.columns), zap.Any("colInfos", input.colInfos))
+		log.Error("avro: converting value to Avro binary failed", zap.Error(err))
 		return nil, cerror.WrapError(cerror.ErrAvroEncodeToBinary, err)
 	}
 
@@ -218,13 +214,13 @@ func (a *BatchEncoder) AppendRowChangedEvent(
 
 	key, err := a.encodeKey(ctx, topic, e)
 	if err != nil {
-		log.Error("avro encoding key failed", zap.Error(err))
+		log.Error("avro encoding key failed", zap.Error(err), zap.Any("event", e))
 		return errors.Trace(err)
 	}
 
 	value, err := a.encodeValue(ctx, topic, e)
 	if err != nil {
-		log.Error("avro encoding value failed", zap.Error(err))
+		log.Error("avro encoding value failed", zap.Error(err), zap.Any("event", e))
 		return errors.Trace(err)
 	}
 

--- a/pkg/sink/codec/avro/decoder_test.go
+++ b/pkg/sink/codec/avro/decoder_test.go
@@ -20,11 +20,45 @@ import (
 	"time"
 
 	timodel "github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tiflow/cdc/entry"
 	"github.com/pingcap/tiflow/cdc/model"
 	"github.com/pingcap/tiflow/pkg/config"
 	"github.com/pingcap/tiflow/pkg/sink/codec/common"
 	"github.com/stretchr/testify/require"
 )
+
+func TestDMLEventE2E(t *testing.T) {
+	helper := entry.NewSchemaTestHelper(t)
+	defer helper.Close()
+
+	helper.Tk().MustExec("use test")
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	codecConfig := common.NewConfig(config.ProtocolAvro)
+	codecConfig.EnableTiDBExtension = true
+	encoder, err := SetupEncoderAndSchemaRegistry4Testing(ctx, codecConfig)
+	defer TeardownEncoderAndSchemaRegistry4Testing()
+	require.NoError(t, err)
+	require.NotNil(t, encoder)
+
+	_ = helper.DDL2Event(`create table t(a varchar(64) not null, b varchar(64) default null, primary key(a))`)
+
+	event := helper.DML2Event(`insert into t values('a', 'b')`, "test", "t")
+
+	topic := "avro-test-topic"
+	err = encoder.AppendRowChangedEvent(ctx, topic, event, func() {})
+	require.NoError(t, err)
+
+	event = helper.DML2Event(`insert into t(a) values ('b')`, "test", "t")
+	err = encoder.AppendRowChangedEvent(ctx, topic, event, func() {})
+	require.NoError(t, err)
+
+	event = helper.DML2Event(`insert into t(a) values ('')`, "test", "t")
+	err = encoder.AppendRowChangedEvent(ctx, topic, event, func() {})
+	require.NoError(t, err)
+}
 
 func TestDecodeEvent(t *testing.T) {
 	codecConfig := common.NewConfig(config.ProtocolAvro)

